src/bokeh/io/notebook.py

```diff
@@ -640,6 +653,69 @@ def _server_url(url: str, port: int | None) -> str:
     else:
         return f"http://{url.split(':')[0]}{port_}{'/'}"
 
+
+def _remote_jupyter_proxy_url(port: int | None) -> str:
+    """ Callable to configure Bokeh's show method when a proxy must be
+    configured. If port is None we're asking about the URL
+    for the origin header.
+
+    Taken from documentation here:
+
+       https://docs.bokeh.org/en/latest/docs/user_guide/output/jupyter.html#jupyterhub
+
+    and made an implicit override when JUPYTER_BOKEH_EXTERNAL_URL is defined in
+    a user's environment to the external hostname of the hub, e.g. https://our-hub.edu
+
+    Args:
+       port (int):
+           random port generated by bokeh to avoid re-using recently closed ports
+
+       Returns:
+           str:
+               URL capable of traversing the JupyterHub proxy to return to this notebook session.
+    """
+    base_url = os.environ['JUPYTER_BOKEH_EXTERNAL_URL']
+    host = urllib.parse.urlparse(base_url).netloc
+
+    # If port is None we're asking for the URL origin
+    # so return the public hostname.
+    if port is None:
+        return host
+
+    service_url_path = os.environ['JUPYTERHUB_SERVICE_PREFIX']
+    proxy_url_path = f'proxy/{port}'
+
+    user_url = urllib.parse.urljoin(base_url, service_url_path)
+    full_url = urllib.parse.urljoin(user_url, proxy_url_path)
+    return full_url
+
+
+def _update_notebook_url_from_env(notebook_url: [str | ProxyUrlFunc]) -> [str | ProxyUrlFunc]:
+    """If the environment variable ``JUPYTER_BOKEH_EXTERNAL_URL`` is defined, returns a function which
+    generates URLs which can traverse the JupyterHub proxy.  Otherwise returns ``notebook_url`` unmodified.
+
+    A warning is issued if ``notebook_url`` is not the default ('localhost:8888') and
+    ``JUPYTER_BOKEH_EXTERNAL_URL`` is also defined since setting the environment variable
+    makes specifying ``notebook_url`` irrelevant.
+
+    Args:
+       notebook_url (str | ProxyUrlFunc):
+          Either a URL string which defaults to 'localhost:8888', or a function that given a port
+          number will generate a URL suitable for traversing the JupyterHub proxy.
+
+    Returns:
+       str | ProxyUrlFunc
+          Either a URL string or a function that generates a URL string given a port number.  The
+          latter function may be user supplied as the input parameter or defined internally by Bokeh
+          when ``JUPYER_BOKEH_EXTERNAL_URL`` is set.
+    """
+    if os.environ.get("JUPYTER_BOKEH_EXTERNAL_URL"):
+        if notebook_url != "localhost:8888":
```

Comment:
If we want to check against a default value that is not None, then we will need to define something like
```
DEFAULT_JUPYTER_URL = "localhost:8888"
```
somewhere, and then use that here, and also wherever the defaults are specified, in order to avoid duplication in multiple places (and the possibility that in the future one gets updated but not another)